doc/developer/design/20320328_persist_columnar.md

```diff
@@ -0,0 +1,940 @@
+  // String
+  grantor: RoleId,
+  // u64
+  acl_mode: AclMode,
+}
+```
+
+</td>
+<td>
+
+```
+StructArray<{
+  grantee: String,
+  grantor: String,
+  acl_mode: u64,
+}>
+```
+
+</td>
+<td>
+
+Structure the data as it is in Rust.
+
+> **Alternative:** Encode an MzAclItem into a binary format and store it as a blob.
+
+</td>
+
+<tr>
+<td> AclItem </td>
+<td>
+
+```rust
+struct AclItem {
+  // u32
+  grantee: Oid,
+  // u32
+  grantor: Oid,
+  // u64
+  acl_mode: AclMode,
+}
+```
+
+</td>
+<td>
+
+```
+StructArray<{
+  grantee: u32,
+  grantor: u32,
+  acl_mode: u64,
+}>
+```
+
+</td>
+<td>
+
+Structure the data as it is in Rust.
+
+> **Alternative:** It would be relatively easy to stitch together the three
+values that make up an `AclItem` into a `FixedSizeBinary<16>`, it should even
+sort the same as its Rust counterpart.
+
+</td>
+
+<tr>
+<td> Int2Vector </td>
+<td>
+
+```rust
+Vec<i16>
+```
+
+</td>
+<td>
+
+`VariableSizeList<i16>`
+
+</td>
+<td>
+
+Structure the data as it is in Rust.
+
+</td>
+
+</table>
+
+### Extension Types
+
+Arrow has the concept of [Extension Types](https://arrow.apache.org/docs/format/Columnar.html#extension-types)
+which do not change the physical layout of a column, but allow you to identify
+the kind of data stored in a column by tagging it with metadata. At the time
+of writing there are only two accepted [canonical extension types](https://arrow.apache.org/docs/format/CanonicalExtensions.html#official-list)
+but we are free to create our own.
+
+As part of our mapping from `Datum` to Arrow column we could include an
+extension type in the column's metadata, specifically:
+```
+'ARROW:extension:name': 'materialize.persist.<version>.<datum_name>'
+```
+To start `<version>` will just be `1`, but could be used in the future to
+evolve how we represent `Datum`s in Arrow. And we include the 'persist'
+namespacing to make sure we don't collide with any other Materialize Arrow
+formats, e.g. for `COPY TO ...`.
+
+> **Note:** The [guidance](https://arrow.apache.org/docs/format/Columnar.html#extension-types)
+provided by the Arrow Project is to name extension types with a prefix to
+prevent collisions with other applications. I don't think this really matters
+for Persist, but I don't see a reason not to follow this pattern.
+
+### Schemas and Column Naming
+
+When writing Arrow columns as Parquet you need to include a schema, which
+requires each column to be named. Instead of using the externally visible
+column names, we could name the Arrow fields with their column index. For
+example, a table with two columns 'foo' and 'bar' would be persisted with
+column names '0' and '1'. This is a bit forward looking, but aims to support
+two goals:
+
+1. Dropping and adding columns of the same name. We would use an
+   auto-incrementing tag number instead of column index, but would allow us to
+   distinguish between these two states.
+2. Users that consider column names to be sensitive. Definitely not a blocker
```

Comment:
It is :D

This is a very small point but it only comes to mind because schemas being sensitive came up in a recent customer call. But as an example of how this could be beneficial, imagine a debugging tool which prints blob metadata or thinking about storing schemas somewhere in CRDB, now those systems don't have "customer data" so it's possible they could be more widely used.